### ISSUE TITLE
A monad for writing method bodies

### ISSUE THREAD

**Issue by dmcclean** [tomlokhorst/language-cil#6]
The goal of making use of the library read as much like the resulting IL as possible, together with some issues around alpha-conversion, led me to this.

Instead of writing a method body as a [MethodDecl], I propose something like this change from:

```
ioAge :: MethodDef
ioAge = Method [MaStatic, MaPublic] Void "ioAge" []
[ maxStack 11
  , localsInit
  [ Local Int32 "x"
  , Local (ValueType "mscorlib" "System.DateTime") "d1"
  , Local (ValueType "mscorlib" "System.DateTime") "d2"
  ]
, ldstr "What year were you born?"
, call [] Void "mscorlib" "System.Console" "WriteLine" [String]
, call [] String "mscorlib" "System.Console" "ReadLine" []
, call [] Int32 "" "int32" "Parse" [String]
, stloc 0
, call [] (ValueType "mscorlib" "System.DateTime") "mscorlib" "System.DateTime" "get_Now" []
, stloc 1
, ldloca 1
, ldloc 0
, neg
, call [CcInstance] (ValueType "mscorlib" "System.DateTime") "mscorlib" "System.DateTime" "AddYears" [Int32]
, stloc 2
, ldstr "This year, you turn {0}."
, ldloca 2
, call [CcInstance] Int32 "mscorlib" "System.DateTime" "get_Year" []
, box Int32
, call [] Void "mscorlib" "System.Console" "WriteLine" [String, Object]
, ret
]
```

to:

```
dateTime = ValueType "mscorlib" "System.DateTime" -- just for abbreviation

ioAge :: MethodDef
ioAge = Method [MaStatic, MaPublic] Void "ioAge" []
$ do
      maxStack 11 -- this stays for now until the analysis to do it is done
      -- you could keep the same localsInit expression if you wanted to, this is just to demonstrate how it would be done in the case where the code generator is not in a position to assign all the names at once or isn't concerned with names

      x <- freshLocal Int32
      d1 <- freshLocal (dateTime)
      d2 <- freshLocal (dateTime)

      ldstr "What year were you born?"
      call [] Void "mscorlib" "System.Console" "WriteLine" [String]
      call [] String "mscorlib" "System.Console" "ReadLine" []
      call [] Int32 "" "int32" "Parse" [String]
      stloc x
      call [] dateTime "mscorlib" "System.DateTime" "get_Now" []
      stloc d1
      ldloca d1
      ldloc x
      neg
      call [CcInstance] dateTime "mscorlib" "System.DateTime" "AddYears" [Int32]
      stloc d2
      ldstr "This year, you turn {0}."
      ldloca d2
      call [CcInstance] Int32 "mscorlib" "System.DateTime" "get_Year" []
      box Int32
      call [] Void "mscorlib" "System.Console" "WriteLine" [String, Object]
      ret
```

With appropriate supporting code, sketched below:

```
data MethodBuilderState = MState { instructions :: [MethodDecl], nextLocal :: Int,  ... } -- I have a few ideas about the ..., omitted for simplicity

import Control.Monad.State

type MethodBuilder a = State MethodBuilderState a

initialState :: MethodBuilderState
initialState = MState { instructions = [], nextLocal = 0, ... }

buildMethod :: MethodBuilder a -> [MethodDecl]
buildMethod = runState initialState -- not quite, run the maxstack analysis and add the result, add the .localsinit for whatever freshLocals may have been made, etc., but you get the idea

freshLocal :: MethodBuilder Offset -- or could be a name with a possible unique-ifying suffix, or could be one variant for each
freshLocal = do
                      result <- getNextLocal
                      modify (\s -> s { nextLocal = result + 1 })
                      return result

append :: MethodDecl -> MethodBuilder ()
append instr = modify (\s -> s { instructions = (instructions s) ++ [instr] }) -- obviously it would be better for performance to store a backwards list of instructions since we only snoc and never cons, presented forwards to get the idea across
```

and with appropriate changes to the building methods, e.g. from:

```
bgt :: Label -> MethodDecl
bgt = mdecl . Bgt
```

to:

```
bot :: Label -> MethodBuilder ()
bot = append $ mdecl . Bgt
```

(Potentially for customizability you might want to define MethodBuilder as a class instead of a type, make append a function in that class, change the builder type of the example from Label -> MethodBuilder () to (MethodBuilder m) => Label -> m (), etc. This would allow people to use a more feature-rich method building monad if their backend had some reason to do so. Ignore that complexity for now.)

tl;dr The monadic version both looks a lot more like the IL than the [MethodDecl] version and provides a place to hang a few other things that practical code-generators are going to need.
